### ISSUE TITLE
[REVIEW]: matchms - processing and similarity evaluation of mass spectrometry data

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2411]
**Submitting author:** @florian-huber (<a href="http://orcid.org/0000-0002-3535-9406">Florian Huber</a>)
**Repository:** <a href="https://github.com/matchms/matchms" target ="_blank">https://github.com/matchms/matchms</a>
**Version:** 0.4.0
**Editor:** @arfon
**Reviewer:** @bittremieux, @kaibioinfo
**Archive:** <a href="https://doi.org/10.5281/zenodo.4001991" target="_blank">10.5281/zenodo.4001991</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/dadd747809be21db9046445aa0a456d4/status.svg)](https://joss.theoj.org/papers/dadd747809be21db9046445aa0a456d4)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/dadd747809be21db9046445aa0a456d4"><img src="https://joss.theoj.org/papers/dadd747809be21db9046445aa0a456d4/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/dadd747809be21db9046445aa0a456d4/status.svg)](https://joss.theoj.org/papers/dadd747809be21db9046445aa0a456d4)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/matchms/matchms" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@bittremieux & @kaibioinfo, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @arfon know.

✨ **Please try and complete your review in the next six weeks** ✨ 

### Conflict of interest

- [X] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [X] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [X] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/matchms/matchms">repository url</a>?
- [X] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [X] **Contribution and authorship:** Has the submitting author (@florian-huber) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [X] **Installation:** Does installation proceed as outlined in the documentation?
- [X] **Functionality:** Have the functional claims of the software been confirmed?
- [X] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [X] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [X] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [X] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [X] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [X] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [X] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [X] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [X] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [X] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [X] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [X] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/matchms/matchms">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@florian-huber) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @bittremieux, @kaibioinfo it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1038/nbt.3597 is OK
- 10.1073/pnas.1203689109 is OK
- 10.1109/MCSE.2011.37 is OK
- 10.1021/acs.jproteome.8b00717 is OK
- 10.1093/nar/gkz1019 is OK
- 10.1002/jms.1777 is OK

MISSING DOIs

- None

INVALID DOIs

- https://doi.org/10.1038/s41592-019-0686-2 is INVALID because of 'https://doi.org/' prefix
- 10.1021/jasms.8b04453 is INVALID
```

**Comment by arfon** [comment]
@bittremieux, @kaibioinfo - please carry out your review in this issue by updating the checklist above and giving feedback in this issue. The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html

Any questions/concerns please let me know.

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02411/joss.02411/10.21105.joss.02411.pdf)

**Comment by kaibioinfo** [comment]
I added three issues:

One rather major:
Current scope of the software is not clear from the paper (https://github.com/matchms/matchms/issues/95#issue-651066231)

And two minors:
Cosine similarity misses weighting and removal of precursor
https://github.com/matchms/matchms/issues/96#issue-651068504

Regular expression for molecular formulas is ambiguous
https://github.com/matchms/matchms/issues/94#issue-651063454

**Comment by bittremieux** [comment]
I would recommend a few more citations to relevant (Python) packages that cover similar functionality, such as [OpenMS](https://doi.org/10.1038/nmeth.3959)/[pyOpenMS](https://doi.org/10.1002/pmic.201300246) and [spectrum_utils](https://doi.org/10.1021/acs.analchem.9b04884) (disclaimer: this is my own package). [MSnbase](https://doi.org/10.1093/bioinformatics/btr645) is a highly relevant R package as well. Given the high overlap between matchms and [spectrum_utils](https://github.com/bittremieux/spectrum_utils/) a more detailed comparison in terms of spectrum processing functionality and computational efficiency might be warranted, although as spectrum_utils developer I leave the final decision about this to @arfon.

Minor comments for the manuscript:

- Abstract: "full quantitative analysis" -> I suggest to rephrase this to avoid confusion with quantitative metabolomics experiments, i.e. where the abundance of the metabolites is measured.
- The correct reference for Numba is: https://doi.org/10.1145/2833157.2833162

I agree with the issues raised by @kaibioinfo, especially on scope. This is not necessarily a major problem, but should be clarified better. Additionally, I have created two more small issues:

- [matchms/matchms#100](https://github.com/matchms/matchms/issues/100): Install matchms using pip.
- [matchms/matchms#101](https://github.com/matchms/matchms/issues/101): Documentation improvements.

**Comment by arfon** [comment]
Thanks  @bittremieux, @kaibioinfo for your feedback/reviews. @florian-huber - please let us know here when you've managed to incorporate feedback from the reviews.

**Comment by florian-huber** [comment]
Thanks @arfon and in particular @bittremieux and @kaibioinfo for reviewing.
The raised comments were very valuable and helped us improve the package.

We generally agree with the feedback we received, and think that we could address most issues.
The main exception is making matchms a pip installable package (https://github.com/matchms/matchms/issues/100), which is complicated because matchms relies on ``rdkit``, a conda package. Still, we hope to address this in the future, maybe releasing a lightweight matchms version for pip.

Other than that, we incorporated https://github.com/matchms/matchms/issues/94, https://github.com/matchms/matchms/issues/101 and the minor comments by @bittremieux above. I believe we addressed the main points of https://github.com/matchms/matchms/issues/96 by fixing and extending the cosine score functions. In addition, we currently work on providing more filter functions do address the remaining points (PR https://github.com/matchms/matchms/pull/115). Since most of the still missing functionality could -though cumbersome- be achieve using present filters, this might be less urgent with respect to the JOSS manuscript.

Finally, the major point raised by @kaibioinfo in https://github.com/matchms/matchms/issues/95 (overlapping with recommendation by @bittremieux  to make the scope clearer) was addressed in two ways.
We extended the importing options of matchms, now including mzML, mzXML, msp, and metabolomics-USI. In addition, we now refer to more related packages and state more clearly the limitations of matchms (see https://github.com/matchms/matchms/issues/95 for more detailed description).

### Manuscript and code update:
- the mentioned changes were made into a new matchms release (version 0.5.0).
- the edited paper and bibliography was added to the master branch.
- we added one additional author (Efraín Manuel Villanueva Castilla, who contributed code to this latest release). 

Please let us know if that is sufficiently addressing the feedback, or further changes need to be made.
Many thanks!

**Comment by kaibioinfo** [comment]
I'm happy with the revision.

**Comment by florian-huber** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02411/joss.02411/10.21105.joss.02411.pdf)

**Comment by bittremieux** [comment]
It looks good for me as well, I just have a few really minor comments:

- Missing reference for mzML on page 1: [10.1074/mcp.R110.000133](http://www.mcponline.org/content/10/1/R110.000133.abstract)
- I would recommend consistent and correct spelling of some terms throughout the manuscript and figures, for example: Inchi/InchiKey -> InChI/InChIKey, Smiles -> SMILES, _m/z_ should be italicized, NumPy, SciPy

**Comment by florian-huber** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02411/joss.02411/10.21105.joss.02411.pdf)

**Comment by florian-huber** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02411/joss.02411/10.21105.joss.02411.pdf)

**Comment by florian-huber** [comment]
Thanks for the comments @bittremieux.
We changed spelling as you suggested (in text and figures).

**Comment by bittremieux** [comment]
Excellent. I don't think the mzML paper is actually cited yet in the manuscript where it is first mentioned though.

**Comment by florian-huber** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02411/joss.02411/10.21105.joss.02411.pdf)

**Comment by florian-huber** [comment]
Thanks a lot for checking this! I corrected the reference entry, it should now be inserted where mzML is first mentioned.

**Comment by florian-huber** [comment]
@arfon Let me know if there is anything we should do to proceed. Thanks!

**Comment by arfon** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02411/joss.02411/10.21105.joss.02411.pdf)

**Comment by arfon** [comment]
@bittremieux & @kaibioinfo - if you have a moment to revisit the checklists above to mark off the last few items (which I believe have now been addressed) that would be much appreciated.

**Comment by arfon** [comment]
@florian-huber - I've made a couple of minor edits in https://github.com/matchms/matchms/pull/137.

After merging https://github.com/matchms/matchms/pull/137, could you make a new release of this software that includes the changes that have resulted from this review. Then, please make an archive of the software in Zenodo/figshare/other service and update this thread with the DOI of the archive? For the Zenodo/figshare archive, please make sure that:

- The title of the archive is the same as the JOSS paper title
- That the authors of the archive are the same as the JOSS paper authors

I can then move forward with accepting the submission.

**Comment by florian-huber** [comment]
I merged the minor edits and made a new release (0.5.2) which is now on Zenodo. 
The DOI is 10.5281/zenodo.4001991

**Comment by bittremieux** [comment]
Checklist completed.

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.4001991 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4001991" target="_blank">10.5281/zenodo.4001991</a> is the archive.

**Comment by arfon** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1038/nbt.3597 is OK
- 10.1145/2833157.2833162 is OK
- 10.1073/pnas.1203689109 is OK
- 10.1038/s41592-019-0686-2 is OK
- 10.1109/MCSE.2011.37 is OK
- 10.1021/acs.jproteome.8b00717 is OK
- 10.1093/nar/gkz1019 is OK
- 10.1002/jms.1777 is OK
- 10.1002/pmic.201300246 is OK
- 10.1038/nmeth.3959 is OK
- 10.1021/acs.analchem.9b04884 is OK
- 10.1093/bioinformatics/bty046 is OK
- 10.1101/2020.05.09.086066 is OK
- 10.1074/mcp.R110.000133 is OK

MISSING DOIs

- None

INVALID DOIs

- 10.1021/jasms.8b04453 is INVALID
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1680

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1680, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
> ```
> INVALID DOIs
> 
> - 10.1021/jasms.8b04453 is INVALID
```

Not sure what's going on with that DOI, (perhaps the publisher never registered it) but it's definitely correct for this: https://pubs.acs.org/doi/pdf/10.1021/jasms.8b04453

@florian-huber - this is ready to accept and publish. The EiC on rotation will handle this in the next couple of days.

**Comment by florian-huber** [comment]
Thanks a lot @arfon!

And short feedback: as first-time submitter to JOSS I really liked the clear and transparent process, as well as the constructive feedback. Really nice!

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1300325537156657152) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1685
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02411](https://doi.org/10.21105/joss.02411)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by arfon** [comment]
@bittremieux, @kaibioinfo - many thanks for your reviews here ✨ 

@florian-huber - your paper is now accepted into JOSS :zap::rocket::boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02411/status.svg)](https://doi.org/10.21105/joss.02411)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02411">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02411/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02411/status.svg
   :target: https://doi.org/10.21105/joss.02411
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02411/status.svg)](https://doi.org/10.21105/joss.02411)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
